phd-tests/tests/src/stats.rs

```diff
@@ -0,0 +1,415 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::time::Duration;
+
+use phd_framework::test_vm::{FakeOximeterSampler, MetricsLocation};
+
+use chrono::{DateTime, Utc};
+use oximeter::types::{ProducerResults, ProducerResultsItem, Sample};
+use oximeter::{Datum, FieldValue};
+use phd_testcase::*;
+use tracing::{trace, warn};
+
+// For convenience when comparing times below.
+const NANOS_PER_SEC: f64 = 1_000_000_000.0;
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, strum::EnumString)]
+#[strum(serialize_all = "snake_case")]
+enum VcpuState {
+    Emulation,
+    Run,
+    Idle,
+    Waiting,
+}
+
+#[derive(Default, Debug)]
+struct VcpuUsageMetric {
+    metrics: HashMap<VcpuState, u64>,
+}
+
+/// A collection of the stats produced by `propolis-server`'s Oximeter producer.
+///
+/// Oximeter producers produce a series of lists of samples, where each list
+/// of samples is conceptually distinct but may still be interesting to
+/// tset. In `propolis-server`, the first list of samples will be
```

Comment:
```suggestion
/// test. In `propolis-server`, the first list of samples will be
```

Reply:
agh yeah, fixed in e42be6010125c87d4443f1d9918ff233e3e40674